src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryTaskLauncher.java

```diff
@@ -0,0 +1,607 @@
+     *
+     * @param packageId id of the package to upload the bits to
+     * @param request the {@link AppDeploymentRequest} with information about the location
+     *                of the bits
+     * @return packageId
+     */
+    protected Mono<String> uploadPackage(String packageId, AppDeploymentRequest request) {
+
+        try {
+            return client.packages()
+                .upload(UploadPackageRequest.builder()
+                    .packageId(packageId)
+                    .bits(request.getResource().getInputStream())
+                    .build())
+                .log("stream.uploadPackage")
+                .map(Package::getId)
+                .log("stream.uploadedPackageId");
+        } catch (IOException e) {
+            return Mono.error(e);
+        }
+    }
+
+    private Set<String> servicesToBind(AppDeploymentRequest request) {
+        Set<String> services = new HashSet<>();
+        services.addAll(properties.getServices());
+        services.addAll(commaDelimitedListToSet(request.getDeploymentProperties().get(SERVICES_PROPERTY_KEY)));
+
+        return services;
+    }
+
+    /**
+     * Create a new {@link Droplet} based upon packageId.
+     *
+     * @param packageId The package id the droplet is associated with
+     * @return {@link Mono} containing the {@link Droplet}'s ID.
+     */
+    private Mono<String> createDroplet(String packageId, AppDeploymentRequest appDeploymentRequest) {
+        Map<String, String> environmentVariables = new HashMap<>(1);
+
+        try {
+            environmentVariables.put("SPRING_APPLICATION_JSON", new ObjectMapper().writeValueAsString(appDeploymentRequest.getDefinition().getProperties()));
+        }
+        catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+
+        return client.packages()
+            .stage(StagePackageRequest.builder()
+                .packageId(packageId)
+                .stagingDiskInMb(diskQuota(appDeploymentRequest))
+                .stagingMemoryInMb(memory(appDeploymentRequest))
+                .environmentVariables(environmentVariables)
+                .build())
+            .log("stream.stageDroplet")
+            .map(Droplet::getId)
+            .log("stream.dropletId");
+    }
+
+    private int diskQuota(AppDeploymentRequest request) {
+        return parseInt(
+            request.getDeploymentProperties().getOrDefault(DISK_PROPERTY_KEY, valueOf(properties.getDisk())));
+    }
+
+    private int memory(AppDeploymentRequest request) {
+        return parseInt(
+            request.getDeploymentProperties().getOrDefault(MEMORY_PROPERTY_KEY, valueOf(properties.getMemory())));
+    }
+
+    private TaskStatus mapTaskToStatus(GetTaskResponse getTaskResponse) {
+
+        switch (getTaskResponse.getState()) {
+            case SUCCEEDED_STATE:
+                return new TaskStatus(getTaskResponse.getId(), LaunchState.complete, null);
```

Comment:
not specifically, but anything the CF API has to offer, we could expose, I guess...